node/src/components/consensus/highway_core/state.rs

```diff
@@ -472,16 +474,60 @@ impl<C: Context> State<C> {
         let threshold = self.total_weight() / 2;
         if endorsed > threshold {
             info!(%unit, "Unit endorsed by at least 1/2 of validators.");
-            let mut fully_endorsed = self.incomplete_endorsements.remove(&unit).unwrap();
-            let endorsed_map = self
-                .weights()
-                .keys()
-                .map(|vidx| fully_endorsed.remove(&vidx))
-                .collect();
-            self.endorsements.insert(unit, endorsed_map);
-            // When unit gets endorsed, it becomes safe to cite.
-            let creator = self.unit(&unit).creator;
-            self.panoramas.endorsed(creator, unit);
+            self.endorsed(unit);
+        }
+    }
+
+    /// Updates the state on newly endorsed unit.
+    fn endorsed(&mut self, uhash: C::Hash) {
+        let mut fully_endorsed = self.incomplete_endorsements.remove(&uhash).unwrap();
+        let endorsed_map = self
+            .weights()
+            .keys()
+            .map(|vidx| fully_endorsed.remove(&vidx))
+            .collect();
+        self.endorsements.insert(uhash, endorsed_map);
+        // When unit gets endorsed, it becomes safe to cite.
+        let creator = self.unit(&uhash).creator;
+
+        if self.citable_panorama().next_seq_num(self, creator) >= self.unit(&uhash).seq_number {
+            // Previously endorsed unit comes later than `uhash`.
```

Comment:
> Note that the previous unit isn't necessarily endorsed, too.

Isn't it? If I endorse your unit _u2_ then it implicitly endorses your previous unit _u1_.

>  It might not cite the equivocator at all, or be cited by another endorsed unit.

Maybe, but if later unit is endorsed then it means there's no faults visible in the unit's past since _u2_ cites _u1_ and includes all of its past as well.

Reply:
Sure, nothing wrong with the code. It's just that the comment doesn't match the definition of "endorsed" as in the paper, which just doesn't _call_ implicitly endorsed units "endorsed".

Reply:
Done in [`361bdd6` (#572)](https://github.com/CasperLabs/casper-node/pull/572/commits/361bdd69ea987f47669f8a3f482a76c986d78b63)